File: src/PokerHub.Client/Services/IndexedDbService.cs
using Microsoft.JSInterop;
using PokerHub.Application.DTOs.Tournament;
using System.Text.Json;

namespace PokerHub.Client.Services;

/// <summary>
/// Wrapper C# para o window.pokerhubDB (indexeddb-service.js).
/// Todas as operações têm try/catch — IndexedDB pode falhar em modo incógnito.
/// </summary>
public class IndexedDbService(IJSRuntime js)
{
    private readonly IJSRuntime _js = js;

    private static readonly JsonSerializerOptions _jsonOpts = new()
    {
        PropertyNameCaseInsensitive = true
    };

    // ── Inicialização ────────────────────────────────────────────────────────

    /// <summary>Pré-aquece a conexão com o IndexedDB.</summary>
    public async Task<bool> InitializeAsync()
    {
        try   { return await _js.InvokeAsync<bool>("pokerhubDB.initialize"); }
        catch { return false; }
    }

    // ── Timer State ──────────────────────────────────────────────────────────

    public async Task SaveTimerStateAsync(Guid tournamentId, TimerStateDto state)
    {
        try { await _js.InvokeVoidAsync("pokerhubDB.saveTimerState", tournamentId.ToString(), state); }
        catch { /* offline fallback silencioso */ }
    }

    public async Task<TimerStateDto?> GetTimerStateAsync(Guid tournamentId)
    {
        try { return await _js.InvokeAsync<TimerStateDto?>("pokerhubDB.getTimerState", tournamentId.ToString()); }
        catch { return null; }
    }

    public async Task DeleteTimerStateAsync(Guid tournamentId)
    {
        try { await _js.InvokeVoidAsync("pokerhubDB.deleteTimerState", tournamentId.ToString()); }
        catch { }
    }

    // ── Active Tournament ────────────────────────────────────────────────────

    public async Task SaveActiveTournamentAsync(TournamentDetailDto tournament)
    {
        try { await _js.InvokeVoidAsync("pokerhubDB.saveActiveTournament", tournament); }
        catch { }
    }

    public async Task<TournamentDetailDto?> GetActiveTournamentAsync(Guid tournamentId)
    {
        try { return await _js.InvokeAsync<TournamentDetailDto?>("pokerhubDB.getActiveTournament", tournamentId.ToString()); }
        catch { return null; }
    }

    public async Task DeleteActiveTournamentAsync(Guid tournamentId)
    {
        try { await _js.InvokeVoidAsync("pokerhubDB.deleteActiveTournament", tournamentId.ToString()); }
        catch { }
    }

    // ── Blind Structure ──────────────────────────────────────────────────────

    public async Task SaveBlindStructureAsync(Guid tournamentId, IReadOnlyList<BlindLevelDto> blinds)
    {
        try { await _js.InvokeVoidAsync("pokerhubDB.saveBlindStructure", tournamentId.ToString(), blinds); }
        catch { }
    }

    public async Task<List<BlindLevelDto>?> GetBlindStructureAsync(Guid tournamentId)
    {
        try { return await _js.InvokeAsync<List<BlindLevelDto>?>("pokerhubDB.getBlindStructure", tournamentId.ToString()); }
        catch { return null; }
    }

    // ── Offline Queue ────────────────────────────────────────────────────────

    public async Task EnqueueActionAsync(OfflineAction action)
    {
        try { await _js.InvokeVoidAsync("pokerhubDB.enqueueOfflineAction", action); }
        catch { }
    }

    /// <summary>
    /// Helper para criar e enfileirar uma ação a partir do tipo e payload.
    /// </summary>
    public Task EnqueueActionAsync(string actionType, object payload)
        => EnqueueActionAsync(OfflineAction.Create(actionType, payload));

    public async Task<OfflineAction?> DequeueActionAsync()
    {
        try { return await _js.InvokeAsync<OfflineAction?>("pokerhubDB.dequeueOfflineAction"); }
        catch { return null; }
    }

    public async Task<List<OfflineAction>> PeekQueueAsync()
    {
        try { return await _js.InvokeAsync<List<OfflineAction>>("pokerhubDB.peekOfflineQueue") ?? []; }
        catch { return []; }
    }

    public async Task<int> GetQueueSizeAsync()
    {
        try { return await _js.InvokeAsync<int>("pokerhubDB.getOfflineQueueSize"); }
        catch { return 0; }
    }

    public async Task ClearQueueAsync()
    {
        try { await _js.InvokeVoidAsync("pokerhubDB.clearOfflineQueue"); }
        catch { }
    }

    public async Task IncrementRetriesAsync(string actionId)
    {
        try { await _js.InvokeVoidAsync("pokerhubDB.incrementRetries", actionId); }
        catch { }
    }

    // ── User Session ─────────────────────────────────────────────────────────

    public async Task SaveUserSessionAsync(object session)
    {
        try { await _js.InvokeVoidAsync("pokerhubDB.saveUserSession", session); }
        catch { }
    }

    public async Task<T?> GetUserSessionAsync<T>()
    {
        try { return await _js.InvokeAsync<T?>("pokerhubDB.getUserSession"); }
        catch { return default; }
    }

    public async Task ClearUserSessionAsync()
    {
        try { await _js.InvokeVoidAsync("pokerhubDB.clearUserSession"); }
        catch { }
    }

    // ── Rankings Cache ────────────────────────────────────────────────────────

    public async Task SaveRankingsCacheAsync(Guid leagueId, object data)
    {
        try { await _js.InvokeVoidAsync("pokerhubDB.saveRankingsCache", leagueId.ToString(), data); }
        catch { }
    }

    public async Task<T?> GetRankingsCacheAsync<T>(Guid leagueId)
    {
        try { return await _js.InvokeAsync<T?>("pokerhubDB.getRankingsCache", leagueId.ToString()); }
        catch { return default; }
    }

    // ── League Data Cache ─────────────────────────────────────────────────────

    public async Task SaveLeagueDataAsync(Guid leagueId, object data)
    {
        try { await _js.InvokeVoidAsync("pokerhubDB.saveLeagueData", leagueId.ToString(), data); }
        catch { }
    }

    public async Task<T?> GetLeagueDataAsync<T>(Guid leagueId)
    {
        try { return await _js.InvokeAsync<T?>("pokerhubDB.getLeagueData", leagueId.ToString()); }
        catch { return default; }
    }

    // ── Utilitários ───────────────────────────────────────────────────────────

    /// <summary>Limpa todos os stores — usar no logout.</summary>
    public async Task ClearAllAsync()
    {
        try { await _js.InvokeVoidAsync("pokerhubDB.clearAll"); }
        catch { }
    }

    /// <summary>
    /// Deserializa o PayloadJson de um OfflineAction para o tipo desejado.
    /// </summary>
    public static T? DeserializePayload<T>(OfflineAction action)
    {
        try   { return JsonSerializer.Deserialize<T>(action.PayloadJson, _jsonOpts); }
        catch { return default; }
    }
}

File: src/PokerHub.Client/wwwroot/js/indexeddb-service.js
// PokerHub IndexedDB Service
// Cache offline para dados de torneio, timer, fila de ações e sessão.
// Todas as operações são wrapped em try/catch — IndexedDB pode falhar em modo incógnito.

'use strict';

(function (window) {
    const DB_NAME    = 'pokerhub-offline';
    const DB_VERSION = 1;

    const STORES = {
        ACTIVE_TOURNAMENT: 'active-tournament',
        TIMER_STATE:       'timer-state',
        BLIND_STRUCTURE:   'blind-structure',
        OFFLINE_QUEUE:     'offline-queue',
        USER_SESSION:      'user-session',
        RANKINGS_CACHE:    'rankings-cache',
        LEAGUE_DATA:       'league-data'
    };

    // TTL em ms
    const TTL = {
        RANKINGS: 24 * 60 * 60 * 1000,  // 24h
        LEAGUE:    1 * 60 * 60 * 1000   //  1h
    };

    let _db = null;

    // ── Conexão ─────────────────────────────────────────────────────────────

    function openDB() {
        if (_db) return Promise.resolve(_db);

        return new Promise((resolve, reject) => {
            const req = indexedDB.open(DB_NAME, DB_VERSION);

            req.onupgradeneeded = e => {
                const db = e.target.result;
                Object.values(STORES).forEach(storeName => {
                    if (!db.objectStoreNames.contains(storeName)) {
                        const store = db.createObjectStore(storeName, { keyPath: 'id' });
                        if (storeName === STORES.OFFLINE_QUEUE) {
                            store.createIndex('timestamp', 'timestamp');
                        }
                    }
                });
            };

            req.onsuccess = e => {
                _db = e.target.result;

                // Reconexão automática se a conexão for fechada inesperadamente
                _db.onclose = () => { _db = null; };
                _db.onversionchange = () => { _db.close(); _db = null; };

                resolve(_db);
            };

            req.onerror   = e => reject(e.target.error);
            req.onblocked = () => reject(new Error('IndexedDB bloqueado por outra aba'));
        });
    }

    // ── CRUD genérico ────────────────────────────────────────────────────────

    async function get(storeName, id) {
        const db = await openDB();
        return new Promise((resolve, reject) => {
            const req = db.transaction(storeName, 'readonly')
                          .objectStore(storeName)
                          .get(id);
            req.onsuccess = e => resolve(e.target.result ?? null);
            req.onerror   = e => reject(e.target.error);
        });
    }

    async function put(storeName, data) {
        const db = await openDB();
        return new Promise((resolve, reject) => {
            const req = db.transaction(storeName, 'readwrite')
                          .objectStore(storeName)
                          .put(data);
            req.onsuccess = () => resolve();
            req.onerror   = e => reject(e.target.error);
        });
    }

    async function remove(storeName, id) {
        const db = await openDB();
        return new Promise((resolve, reject) => {
            const req = db.transaction(storeName, 'readwrite')
                          .objectStore(storeName)
                          .delete(id);
            req.onsuccess = () => resolve();
            req.onerror   = e => reject(e.target.error);
        });
    }

    async function getAll(storeName) {
        const db = await openDB();
        return new Promise((resolve, reject) => {
            const req = db.transaction(storeName, 'readonly')
                          .objectStore(storeName)
                          .getAll();
            req.onsuccess = e => resolve(e.target.result ?? []);
            req.onerror   = e => reject(e.target.error);
        });
    }

    async function clear(storeName) {
        const db = await openDB();
        return new Promise((resolve, reject) => {
            const req = db.transaction(storeName, 'readwrite')
                          .objectStore(storeName)
                          .clear();
            req.onsuccess = () => resolve();
            req.onerror   = e => reject(e.target.error);
        });
    }

    async function count(storeName) {
        const db = await openDB();
        return new Promise((resolve, reject) => {
            const req = db.transaction(storeName, 'readonly')
                          .objectStore(storeName)
                          .count();
            req.onsuccess = e => resolve(e.target.result);
            req.onerror   = e => reject(e.target.error);
        });
    }

    // ── API pública ──────────────────────────────────────────────────────────

    window.pokerhubDB = {

        // ── CRUD genérico (exposto para casos avançados) ─────────────────────

        async get(store, id) {
            try { return await get(store, id); }
            catch (e) { console.warn('[pokerhubDB] get error:', e); return null; }
        },

        async put(store, data) {
            try { await put(store, data); }
            catch (e) { console.warn('[pokerhubDB] put error:', e); }
        },

        async delete(store, id) {
            try { await remove(store, id); }
            catch (e) { console.warn('[pokerhubDB] delete error:', e); }
        },

        async getAll(store) {
            try { return await getAll(store); }
            catch (e) { console.warn('[pokerhubDB] getAll error:', e); return []; }
        },

        async clear(store) {
            try { await clear(store); }
            catch (e) { console.warn('[pokerhubDB] clear error:', e); }
        },

        // ── Timer State ──────────────────────────────────────────────────────

        async saveTimerState(tournamentId, state) {
            try {
                await put(STORES.TIMER_STATE, { id: tournamentId, ...state, savedAt: Date.now() });
            } catch (e) { console.warn('[pokerhubDB] saveTimerState error:', e); }
        },

        async getTimerState(tournamentId) {
            try { return await get(STORES.TIMER_STATE, tournamentId); }
            catch (e) { console.warn('[pokerhubDB] getTimerState error:', e); return null; }
        },

        async deleteTimerState(tournamentId) {
            try { await remove(STORES.TIMER_STATE, tournamentId); }
            catch (e) { console.warn('[pokerhubDB] deleteTimerState error:', e); }
        },

        // ── Active Tournament ────────────────────────────────────────────────

        async saveActiveTournament(tournament) {
            try {
                await put(STORES.ACTIVE_TOURNAMENT, { ...tournament, savedAt: Date.now() });
            } catch (e) { console.warn('[pokerhubDB] saveActiveTournament error:', e); }
        },

        async getActiveTournament(tournamentId) {
            try { return await get(STORES.ACTIVE_TOURNAMENT, tournamentId); }
            catch (e) { console.warn('[pokerhubDB] getActiveTournament error:', e); return null; }
        },

        async deleteActiveTournament(tournamentId) {
            try { await remove(STORES.ACTIVE_TOURNAMENT, tournamentId); }
            catch (e) { console.warn('[pokerhubDB] deleteActiveTournament error:', e); }
        },

        // ── Blind Structure ──────────────────────────────────────────────────

        async saveBlindStructure(tournamentId, blinds) {
            try {
                await put(STORES.BLIND_STRUCTURE, { id: tournamentId, blinds, savedAt: Date.now() });
            } catch (e) { console.warn('[pokerhubDB] saveBlindStructure error:', e); }
        },

        async getBlindStructure(tournamentId) {
            try {
                const row = await get(STORES.BLIND_STRUCTURE, tournamentId);
                return row ? row.blinds : null;
            } catch (e) { console.warn('[pokerhubDB] getBlindStructure error:', e); return null; }
        },

        // ── Offline Queue ────────────────────────────────────────────────────

        async enqueueOfflineAction(action) {
            // action: { id, actionType, payloadJson, timestamp, retries }
            try {
                await put(STORES.OFFLINE_QUEUE, { ...action, timestamp: action.timestamp ?? Date.now() });
            } catch (e) { console.warn('[pokerhubDB] enqueueOfflineAction error:', e); }
        },

        async dequeueOfflineAction() {
            // Retorna a ação mais antiga e a remove da fila
            try {
                const db = await openDB();
                return new Promise((resolve, reject) => {
                    const tx    = db.transaction(STORES.OFFLINE_QUEUE, 'readwrite');
                    const store = tx.objectStore(STORES.OFFLINE_QUEUE);
                    const idx   = store.index('timestamp');

                    const req = idx.openCursor(null, 'next'); // cursor ascendente (mais antigo primeiro)
                    req.onsuccess = e => {
                        const cursor = e.target.result;
                        if (!cursor) { resolve(null); return; }

                        const item = cursor.value;
                        cursor.delete();
                        resolve(item);
                    };
                    req.onerror = e => reject(e.target.error);
                });
            } catch (e) {
                console.warn('[pokerhubDB] dequeueOfflineAction error:', e);
                return null;
            }
        },

        async peekOfflineQueue() {
            // Retorna todas as ações na fila sem remover (para exibição)
            try {
                const db = await openDB();
                return new Promise((resolve, reject) => {
                    const req = db.transaction(STORES.OFFLINE_QUEUE, 'readonly')
                                  .objectStore(STORES.OFFLINE_QUEUE)
                                  .index('timestamp')
                                  .getAll();
                    req.onsuccess = e => resolve(e.target.result ?? []);
                    req.onerror   = e => reject(e.target.error);
                });
            } catch (e) {
                console.warn('[pokerhubDB] peekOfflineQueue error:', e);
                return [];
            }
        },

        async getOfflineQueueSize() {
            try { return await count(STORES.OFFLINE_QUEUE); }
            catch (e) { console.warn('[pokerhubDB] getOfflineQueueSize error:', e); return 0; }
        },

        async clearOfflineQueue() {
            try { await clear(STORES.OFFLINE_QUEUE); }
            catch (e) { console.warn('[pokerhubDB] clearOfflineQueue error:', e); }
        },

        async incrementRetries(actionId) {
            try {
                const row = await get(STORES.OFFLINE_QUEUE, actionId);
                if (row) {
                    row.retries = (row.retries ?? 0) + 1;
                    await put(STORES.OFFLINE_QUEUE, row);
                }
            } catch (e) { console.warn('[pokerhubDB] incrementRetries error:', e); }
        },

        // ── User Session ─────────────────────────────────────────────────────

        async saveUserSession(session) {
            // session: { id: 'session', userId, userName, token, leagueIds, ... }
            try { await put(STORES.USER_SESSION, { id: 'session', ...session }); }
            catch (e) { console.warn('[pokerhubDB] saveUserSession error:', e); }
        },

        async getUserSession() {
            try { return await get(STORES.USER_SESSION, 'session'); }
            catch (e) { console.warn('[pokerhubDB] getUserSession error:', e); return null; }
        },

        async clearUserSession() {
            try { await remove(STORES.USER_SESSION, 'session'); }
            catch (e) { console.warn('[pokerhubDB] clearUserSession error:', e); }
        },

        // ── Rankings Cache ────────────────────────────────────────────────────

        async saveRankingsCache(leagueId, data) {
            try {
                await put(STORES.RANKINGS_CACHE, { id: leagueId, data, savedAt: Date.now() });
            } catch (e) { console.warn('[pokerhubDB] saveRankingsCache error:', e); }
        },

        async getRankingsCache(leagueId) {
            try {
                const row = await get(STORES.RANKINGS_CACHE, leagueId);
                if (!row) return null;
                if (Date.now() - row.savedAt > TTL.RANKINGS) {
                    await remove(STORES.RANKINGS_CACHE, leagueId); // expirado
                    return null;
                }
                return row.data;
            } catch (e) { console.warn('[pokerhubDB] getRankingsCache error:', e); return null; }
        },

        // ── League Data Cache ─────────────────────────────────────────────────

        async saveLeagueData(leagueId, data) {
            try {
                await put(STORES.LEAGUE_DATA, { id: leagueId, data, savedAt: Date.now() });
            } catch (e) { console.warn('[pokerhubDB] saveLeagueData error:', e); }
        },

        async getLeagueData(leagueId) {
            try {
                const row = await get(STORES.LEAGUE_DATA, leagueId);
                if (!row) return null;
                if (Date.now() - row.savedAt > TTL.LEAGUE) {
                    await remove(STORES.LEAGUE_DATA, leagueId); // expirado
                    return null;
                }
                return row.data;
            } catch (e) { console.warn('[pokerhubDB] getLeagueData error:', e); return null; }
        },

        // ── Utilitários ───────────────────────────────────────────────────────

        /** Limpa TUDO (logout / reset) */
        async clearAll() {
            try {
                await Promise.all(Object.values(STORES).map(s => clear(s)));
            } catch (e) { console.warn('[pokerhubDB] clearAll error:', e); }
        },

        /** Inicializa o DB (útil para pré-aquecer na startup) */
        async initialize() {
            try { await openDB(); return true; }
            catch (e) { console.warn('[pokerhubDB] initialize error:', e); return false; }
        }
    };

})(window);
